rfcs/2022-02-24-11552-dd-agent-style-secret-management.md

```diff
@@ -0,0 +1,129 @@
+path = "/path/to/the/command"
+argument = "--config foo=bar"
+timeout = 5
+```
+
+Overall the behaviour for corner cases should follow what's in place for environmment variable interpolation as this is
+a very close feature.
+
+### Implementation
+
+A secret backend that would lie in `./src/config/secret.rs` and would call the user provided executable for secrets,
+cache secrets to avoid calling the backend for the same key multiple time. It would read the configuration file once to
+get its config before further processing, ideally env var interpolating should be supported (this should not be a
+problem). In `./src/config/builder.rs`, `load_builder_from_paths` will still be returning a complete configuration with
+placeholders replaced by secrets.
+
+The `ConfigBuilder` struct will get a new `secret_backend` field (type to something like `SecretBackend`), this means
+that `load_builder_from_paths` will then assert if this field is present before returning to the caller, and if it, the
+config will be reloaded with this `SecretBackend` passed to downstream callee and hook the secret interpolation around
+the same point as [the environment variable interpolation][env-var-hook] it would then query this `SecretBackend` for
+each `ENC[secret_key]`.
+
+The implementation should ease future extension and split the internal API queried by the interpolation logic and the
+secret provider that may see other implementation like: `executable` (the one documented in this RFC), `vault`,
+`k8s-config-map`, `aws secretsmanager`, etc.
+
+## Rationale
+
+- Some users just can't put sensitive information inside their configuration.
+
+## Drawbacks
+
+- Integrating with other third party tools directly like Vault would provide better error management and avoid relying on a custom, user-provided binary.
+- This binary might have to be injected into container images which may be inconvenient, other options like using an external volume may be more acceptable but it would still involve a third party executable and all the associated risks.
+
+## Prior Art
+
+- The Datadog Agent has exactly the same feature, despite its simple approach it works reasonnably well, but it is
+  cannot easily support advanced secret management like certificate distribution/revocation, key rotation, etc.
+- Vault is the standard in the industry, and it comes with all kind of advanced features that cannot really supported
+  by the user-provided executable solution.
+
+## Alternatives
+
+- Integrate with other third party tools: Vault and CSP APIs for secret management to start with.
+- Use the env var interpolation as a secret backend and leverate [K8s ability to expose
+  secret][k8s-env-var-from-secrets] as environment variables, still this would be a good idea to document that as this is a convenient solution avoiding the need to inject a custom binary inside container images. Note that the Datadog Agent is now capable to [do that out-of-the-box][dd-agent-with-k8s-secret].
```

Comment:
We can do this now, right? (cc @spencergilbert) Again, the problem is that some companies do not want to store secrets in the environment.

Reply:
Correct, using something like: https://github.com/vectordotdev/helm-charts/blob/develop/charts/vector/values.yaml#L131-L143 would allow users to put `${AWS_ACCESS_KEY_ID}` in their config and have it interpolated at runtime.

Reply:
Awesome ! I will reflect that this solution is already documented the official chart.

Reply:
It _could_ be better documented I think. We can open an issue to explicitly write it out in the README or include it in the website docs somewhere. I think it's safer to say the tools are there, rather than it's documented 😅